### PR TITLE
cg3: init at version 1.3.7

### PR DESCRIPTION
**Pull request by unhammer**
###### Description of changes

The VISL CG-3 Constraint Grammar parser/compiler/runner. CG is a method for rule-based disambiguation and parsing of natural language. See https://visl.sdu.dk/constraint_grammar.html and https://visl.sdu.dk/cg3.html . 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] (nothing depends on it yet) Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/cg3/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  #   "-DCMAKE_INSTALL_PREFIX=${prefix}/"
+  # ];
+
+  # TODO: tcmalloc unless darwin?
+  # option(ENABLE_PYTHON_BINDINGS "Set to ON to build the Python wrapper" OFF)
+
+  meta = with lib; {
+    homepage = "https://github.com/GrammarSoft/cg3";
+    description = "Constraint Grammar interpreter, compiler and applicator vislcg3";
+    maintainers = [ lib.maintainers ];
```
Comment (viraptor) [inline pkgs/development/interpreters/cg3/default.nix:37]: A copy/paste mistake? This should list actual maintainers.

File: pkgs/development/interpreters/cg3/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    license = licenses.gpl3;
```
Comment (viraptor) [inline pkgs/development/interpreters/cg3/default.nix:38]: Files seem to use the gpl3plus version of the license (https://github.com/GrammarSoft/cg3/blob/990ac64551827508d32f78895a09b1ff331a4c24/src/ApertiumApplicator.cpp#L11)

**Comment by viraptor** [comment]
Congrats on the first PR :)

> What kind of tests are expected here? Should I just add some basic functionality, sanity-checking stuff?

If the package includes tests, they should be run in "checkPhase".
Additionally if there's a way to run a very simple sanity check after installation, it may be added to `passthru.tests` - check the repo for usage examples. It should be completely repeatable, minimal and offline though.

> Should I add release notes? (Is this a "new module" since it's a new package?)

Not for simple packages. That's not needed. Modules are used for configuring nixos systems - they're different than packages.

> Is there a simple way to test on the other architectures when I'm on a Linux machine? :)

ofborg should run those for you if available https://github.com/NixOS/nixpkgs/pull/188011/checks?check_run_id=7973278975
Otherwise you can for example ping `NixOS/darwin-maintainers` group if you'd like someone to validate macos.

File: pkgs/development/interpreters/cg3/default.nix (hunk 1 above)
Comment (unhammer) [reply pkgs/development/interpreters/cg3/default.nix:37]: Oh I wasn't sure about that one. Should I add myself to maintainers/maintainers-list.nix?

File: pkgs/development/interpreters/cg3/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+, icu
+, boost
+}:
+
+stdenv.mkDerivation rec {
+  pname = "cg3";
+  version = "1.3.7";
+
+  src = fetchurl {
+    url = "https://github.com/GrammarSoft/${pname}/releases/download/v${version}/${pname}-${version}.tar.bz2";
```
[the remark this answers is not in the document]
Comment (unhammer) [reply pkgs/development/interpreters/cg3/default.nix:14]: newbie mistake, will fix

File: pkgs/development/interpreters/cg3/default.nix (hunk 2 above)
Comment (unhammer) [reply pkgs/development/interpreters/cg3/default.nix:38]: doh, fixing

**Comment by unhammer** [comment]
> If the package includes unit tests or other post-compile checks, they should be run in "checkPhase".

I tried 

```
+  doCheck = true;
+  checkPhase = ''
+    perl $src/test/runall.pl
+  '';
+
```
but the tests write to files (for diffing) and I got a bunch of errors saying that writing wasn't allowed and that tests failed (but hey the main test script exited with 0 so altogether they passed). Is there a standard way to deal with checkPhase tests that write to files in the test directory?

**Review by viraptor — COMMENTED** [review]
Result of `nixpkgs-review pr 188011` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cg3</li>
  </ul>
</details>

**Comment by viraptor** [comment]
> Is there a standard way to deal with checkPhase tests that write to files in the test directory?

Try to make them write to a temp directory if possible. If not 🤷 we'll have to skip it.

File: pkgs/development/interpreters/cg3/default.nix (hunk 1 above)
Comment (viraptor) [reply pkgs/development/interpreters/cg3/default.nix:45]: Yes, please. The PR should be then 2 commits - one to add yourself, another with all the package changes.

Comment (unhammer) [reply pkgs/development/interpreters/cg3/default.nix]: :heavy_check_mark:

File: pkgs/development/interpreters/cg3/default.nix [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    homepage = "https://github.com/GrammarSoft/cg3";
+    description = "Constraint Grammar interpreter, compiler and applicator vislcg3";
+    maintainers = [ lib.maintainers.unhammer ];
+    license = licenses.gpl3Plus;
+    platforms = platforms.all;
+  };
+};
+
+in
+  cg3
```
Comment (viraptor) [inline pkgs/development/interpreters/cg3/default.nix:61]: This is the same as not using `let ... in`, so you can drop both sides.

File: pkgs/development/interpreters/cg3/default.nix [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+  # TODO, consider optionals:
+  # - Enable tcmalloc unless darwin?
+  # - Enable python bindings?
+
+  meta = with lib; {
+    homepage = "https://github.com/GrammarSoft/cg3";
+    description = "Constraint Grammar interpreter, compiler and applicator vislcg3";
+    maintainers = [ lib.maintainers.unhammer ];
```
Comment (viraptor) [inline pkgs/development/interpreters/cg3/default.nix:54]:
```suggestion
    maintainers = with maintainers; [ unhammer ];
```

**Comment by unhammer** [comment]
> > Is there a standard way to deal with checkPhase tests that write to files in the test directory?
> 
> Try to make them write to a temp directory if possible. If not shrug we'll have to skip it.

Doesn't seem possible without patching the test runners – they also exit 0 no matter how many failures. I'll see if we can do something upstream. But now there's a minimal passthru sanity check at least (and it runs `make check` which also seems to test something).

File: pkgs/development/interpreters/cg3/default.nix (hunk 4 above)
Comment (unhammer) [reply pkgs/development/interpreters/cg3/default.nix:61]: but I refer to cg3 in buildInputs to tests (got an error without that) – is there a better way to get cg3 into test dependencies?

Comment (viraptor) [reply pkgs/development/interpreters/cg3/default.nix:61]: Apologies, I did not notice. That is indeed the way it's used in other packages.

**Review by viraptor — APPROVED** [review]

**Comment by unhammer** [comment]
Awesome, thanks for all the help!

**Pull request merged**